### PR TITLE
NIFI-9969 Add extension restrictions to C2 component definition model

### PR DESCRIPTION
**Pull request by bbende**
# Summary

NIFI-9969 Adds restriction to C2 component definition model and updates the runtime manifest generator to populate the restrictions from the extension manifest.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-manifest/nifi-runtime-manifest-test/src/test/java/org/apache/nifi/runtime/manifest/TestRuntimeManifest.java [hunk 1]
```diff
@@ -139,6 +142,20 @@ public void testRuntimeManifest() throws IOException {
         assertEquals(1, resourceDefinition.getResourceTypes().size());
         assertEquals(ResourceType.FILE, resourceDefinition.getResourceTypes().stream().findFirst().get());
 
+        // Verify FetchHDFS definition has restrictions
```
Comment (bejancsaba) [inline nifi-manifest/nifi-runtime-manifest-test/src/test/java/org/apache/nifi/runtime/manifest/TestRuntimeManifest.java:145]: What do you think about adding assertion for a negative case too? I mean where the ExtensionComponnet is not set to be restricted.

File: nifi-manifest/nifi-runtime-manifest-core/src/main/java/org/apache/nifi/runtime/manifest/impl/StandardRuntimeManifestBuilder.java [hunk 2]
```diff
@@ -348,6 +351,24 @@ private void populateExtensionComponent(final ExtensionManifest extensionManifes
             providedServiceApis.forEach(providedServiceApi -> providedApiTypes.add(createProvidedApiType(providedServiceApi)));
             extensionComponent.setProvidedApiImplementations(providedApiTypes);
         }
+
+        final Restricted restricted = extension.getRestricted();
+        if (restricted != null) {
+            extensionComponent.setRestricted(true);
+            extensionComponent.setRestrictedExplanation(restricted.getGeneralRestrictionExplanation());
+            if (restricted.getRestrictions() != null) {
```
Comment (bejancsaba) [inline nifi-manifest/nifi-runtime-manifest-core/src/main/java/org/apache/nifi/runtime/manifest/impl/StandardRuntimeManifestBuilder.java:359]: It is minor / cosmetic but what do you think about something like this. (Not sure whether functional coding style is preferred or not):
```
Optional.ofNullable(restricted.getRestrictions())
  .map(r -> r.stream().map(this::createRestriction).collect(Collectors.toSet()))
  .ifPresent(restrictions -> extensionComponent.setExplicitRestrictions(restrictions));
```

**Review by bejancsaba — COMMENTED** [review]
+1 from my side (I don't think I can approve anything yet). As I don't have significant knowledge about restrictions I just left two minor comments regarding the code.

File: nifi-manifest/nifi-runtime-manifest-test/src/test/java/org/apache/nifi/runtime/manifest/TestRuntimeManifest.java (hunk 1 above)
Comment (bbende) [reply nifi-manifest/nifi-runtime-manifest-test/src/test/java/org/apache/nifi/runtime/manifest/TestRuntimeManifest.java:150]: Good call, I meant to do that, pushed a new commit

File: nifi-manifest/nifi-runtime-manifest-core/src/main/java/org/apache/nifi/runtime/manifest/impl/StandardRuntimeManifestBuilder.java (hunk 2 above)
Comment (bbende) [reply nifi-manifest/nifi-runtime-manifest-core/src/main/java/org/apache/nifi/runtime/manifest/impl/StandardRuntimeManifestBuilder.java:359]: If everything else in the class was already in this style, then I'd definitely be consistent and do something like that. For now I'd leave it as is.

**Comment by mattyb149** [comment]
+1 LGTM, with the other approvals this is ready to go in. Thanks for the reviews and for the improvement! Merging to main

**Pull request closed without merging**